src/librustc_mir/borrow_check/mod.rs

```diff
@@ -1547,6 +1614,23 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 err.emit();
             },
             Reservation(WriteKind::Mutate) | Write(WriteKind::Mutate) => {
+                match place {
```

Comment:
I think logic is too simplistic. For example, consider this case:

```rust
let mut x = (1, 2);
x.0 = 22;
```

Here, the mutated place will be a field projection, but we still need to record `x` as a "used mut". We want to integrate more with the `is_mutable` code, I think -- in particular, maybe it should return the "root reason" that a path was mutable, even on success (right now it does that only on error). Then, if that root reason is a Local, we can add it to the `used_mut` set? etc